tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +190,371 @@ public void invalidateBufferedData() throws IOException {
             this.lo = pos & BuffMask; // start at BuffSz boundary
-            this.maxHi = this.lo + this.buff.length;
-            if (this.diskPos != this.lo) {
-                super.seek(this.lo); 
-                this.diskPos = this.lo;
-            }
-            int n = this.fillBuffer();
-            this.hi = this.lo + n;
-        }
-	else {
-            // seeking inside current buffer -- no read required
-            if (pos < this.curr) {
-                // if seeking backwards, we must flush to maintain V4
-                this.flushBuffer();
-            }
+            this.fillBuffer();
+            didPageRead = true;
         }
         this.curr = pos;
+        return didPageRead;
     }
-    
-    /* extends this.seek(long) by return if a page has been read (used for statistics) */
-    //TODO come up with better name for seeek %)
-    public boolean seeek(long pos) throws IOException {
-		boolean pageReadNeeded = true;
-		// Assert.check(!this.closed);
-		if (pos >= this.hi || pos < this.lo) {
-			// seeking outside of current buffer -- flush and read
-			this.flushBuffer();
-			this.lo = pos & BuffMask; // start at BuffSz boundary
-			this.maxHi = this.lo + this.buff.length;
-			if (this.diskPos != this.lo) {
-				super.seek(this.lo);
-				this.diskPos = this.lo;
-			}
-			int n = this.fillBuffer();
-			this.hi = this.lo + n;
-		} else {
-			// seeking inside current buffer -- no read required
-			if (pos < this.curr) {
-				// if seeking backwards, we must flush to maintain V4
-				pageReadNeeded = this.flushBuffer();
-			} else {
-				pageReadNeeded = false;
-			}
-		}
-		this.curr = pos;
-		return pageReadNeeded;
-	}
-
-    /* overrides RandomAccessFile.getFilePointer() */
+
+    @Override
     public long getFilePointer() {
         // Assert.check(!this.closed);
         return this.curr;
     }
-    
-    /* overrides RandomAccessFile.length() */
+
+    @Override
     public long length() throws IOException {
         // Assert.check(!this.closed);
-        return Math.max(this.curr, super.length());
+        return this.length;
     }
-    
-    /* overrides RandomAccessFile.read() */
+
+    @Override
+    public void setLength(long newLength) throws IOException {
+        // Per docs:
+        //  > If the present length of the file as returned by the length method is greater than the newLength argument
+        //  > then the file will be truncated. In this case, if the file offset as returned by the getFilePointer
+        //  > method is greater than newLength then after this method returns the offset will be equal to newLength.
+        //
+        // In practice, RandomAccessFile actually moves the pointer if getFilePointer > newLength, regardless of
+        // whether the file was truncated or extended.
+        super.setLength(newLength);
+        if (this.diskPos > newLength) {
+            this.diskPos = super.getFilePointer();
+        }
+
+        if (this.curr > newLength) {
```

Comment:
I first thought that maybe this should just always seek to `diskPos`, but I see this is right, because of the subtle cases around end of file.

I think it's worth explaining this case as for why not:
```
If the buffer landed in a way where the end of the file appears in the middle of the buffer,
such as a short file of just 4 bytes.  When we filled the buffer, the diskPos and curr would
both be in the middle of the buffer, at index 4.  Now consider we seek to the beginning of
the file, but then chop one byte off the file.  In this case, `diskPos` will have moved, but
`curr` should not.

         curr and diskPos
         v
[A B C D * * * * ...]

seek(0)

 curr    diskPos
 v       v
[A B C D * * * * ...]

setLength(3)

 curr  diskPos
 v     v
[A B C * * * * * ...]